lib/railway_ipc/outgoing_message.rb

```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+module RailwayIpc
+  class OutgoingMessage
+    extend Forwardable
+
+    attr_reader :proto, :exchange, :format
+
+    def_delegators :@proto, :uuid, :user_uuid, :correlation_id
+
+    def initialize(proto, exchange, format=nil)
+      proto.uuid = SecureRandom.uuid if proto.uuid.blank?
+      proto.correlation_id = SecureRandom.uuid if proto.correlation_id.blank?
+      @proto = proto
+      @exchange = exchange
+      @format = format
+    end
+
+    def type
+      proto.class.to_s
+    end
+
+    def encoded
+      @encoded ||= encoder.call(self)
+    end
+
+    private
+
+    DEFAULT_ENCODER = RailwayIpc::MessageEncoders::ProtobufBinaryEncoder
+
+    def encoder
+      {
+        'binary_protobuf' => RailwayIpc::MessageEncoders::ProtobufBinaryEncoder,
+        'json_protobuf' => RailwayIpc::MessageEncoders::ProtobufJsonEncoder
+      }.fetch(format, DEFAULT_ENCODER)
```

Comment:
Don't we want to move towards using the Json encoder? If that's the case, could we set the default to that?

Reply:
I don't want to set the as the default just yet. For now I'm making it opt-in where the publisher has to explicitly say it's going to use JSON. This allows us to "upgrade" message formats for each publisher individually, and gives us a way to fall back to the old behavior w/o doing a full re-deploy if something goes wrong. Once _all_ publishers (both in Ruby & Elixir) are emitting JSON, then I'll probably make it the default.